Fixes for main

beckertobias opened the pull request:
SignIn should work now ✅ 

Regarding SignUp:
- I added error handling so we don't have an undhandled error anymore
- But we are still throwing an ApolloError because:

The thing is that I have the following line in my code: `const user = await getTalentById(uid);`. And I don't know in advance if I have a user with that uid or not. So I want to return the user that I get (either `undefined` or an `user`). 

But in the new version of main we do `throw new ApolloError` in the helper function `getTalentById` if we didn't find a user with an ID / UID.

Suggestion: Let's not throw an error in the helper function and instead return talent or undefined.

TobiBecker committed:
add error handling if no user found with id

TobiBecker committed:
fix signin: pass t and change props for InputField

atsteinberg commented:
Sure. Change the graphql if that is necessary for our signup to work. Could you do it for this pr? Or do you want to create a new one?

TobiBecker committed:
return talent or null in getTalentById

beckertobias commented:
Implemented it now as suggested.
Due to the bad internet connection in the train, I fail on testing the signup (just get an error because server cannot be reached - same for graphql playground). 
Need to test that properly before merging.

TobiBecker committed:
fix signup

beckertobias commented:
@atsteinberg  @camifernweh  @pintoedo signup is working now. 
PLease somebody approve. Then main is working at least and we can continue with all the other TO DOs

atsteinberg reviewed (commented):
Hm. I'm not sure I like these changes. From the standpoint of graphql, asking for a talent whose uid does not exist in the db *should* throw an error. Why is it impossible to expect it at this point and deal with it?

My problem is: I've tried to figure out which line actually makes it the case that the 400 response error is logged to the console. Can you help me and walk me through it one of these days? Sorry for being difficult, but there should be an easy way of treating server errors without them showing in the console, or am I missing something here?

TobiBecker commented on apollo/helpers.ts:
```diff
@@ -47,17 +47,13 @@ export const getAllUserIds = (): string[] => {
   } catch (e) {
     handleError(e);
   }
-  if (!talent)
-    throw new ApolloError({
-      errorMessage: `404: No user with id ${id} found in db`,
-    });
```
@atsteinberg this line was throwing the error, if I recall correctly

atsteinberg commented:
yes. Sure. For my taste that is fine for the error to be thrown there.
But why does the error show up in the browser console? I agree that it shouldn’t be there (as a console log to the browser console), since we expect to trigger the graphql error in this case.

> On 25. Feb 2021, at 15:55, TobiBecker <notifications@github.com> wrote:
> 
> 
> @TobiBecker commented on this pull request.
> 
> In apollo/helpers.ts <https://github.com/fullstacksmart/next-kara/pull/29#discussion_r582899007>:
> 
> > -  if (!talent)
> -    throw new ApolloError({
> -      errorMessage: `404: No user with id ${id} found in db`,
> -    });
> @atsteinberg <https://github.com/atsteinberg> this line was throwing the error, if I recall correctly
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/29#pullrequestreview-598652757>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIISTVBFR6TNAPGDGKDLTAZQGLANCNFSM4XSFD3IQ>.
>

TobiBecker commented:
> Hm. I'm not sure I like these changes. From the standpoint of graphql, asking for a talent whose uid does not exist in the db _should_ throw an error. Why is it impossible to expect it at this point and deal with it?
> 

I am dealing with the error. This is not the issue. The issue is that I don't want an error to be thrown each time somebody signs up 😅 
But the signup is working - even with the error.

If you want GraphQL to throw an error in this case, we have to make it different. Then we need one query to check if there is a talent with an uid. And if there is one, we call another query that gets us the query. Not sure if there is a better solution.

atsteinberg commented:
Again, my issue is why the error message shows up in the browser console. Do you know?

> On 25. Feb 2021, at 16:01, TobiBecker <notifications@github.com> wrote:
> 
> 
> Hm. I'm not sure I like these changes. From the standpoint of graphql, asking for a talent whose uid does not exist in the db should throw an error. Why is it impossible to expect it at this point and deal with it?
> 
> I am dealing with the error. This is not the issue. The issue is that I don't want an error to be thrown each time somebody signs up 😅
> But the signup is working - even with the error.
> 
> If you want GraphQL to throw an error in this case, we have to make it different. Then we need one query to check if there is a talent with an uid. And if there is one, we call another query that gets us the query. Not sure if there is a better solution.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/29#issuecomment-785966390>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIIU56AWETTNDJLZX4CLTAZQ4ZANCNFSM4XSFD3IQ>.
>

TobiBecker commented:
ah, just saw your other comment. I can answer later, I'm in a meeting. But I'm not sure why it shows up in the console. At least not out of my head

TobiBecker commented:
yeah sorry, my comment was delayed. Didn't see your other comment when I submitted

TobiBecker commented:
looked into it. But couldn't figure it out yet.
The `throw new ApolloError` makes the error appear in the console of the server - which is fine.
<img width="776" alt="Screenshot 2021-02-25 at 17 01 16" src="https://user-images.githubusercontent.com/50356310/109180490-38adb600-778b-11eb-8e95-6ab523c6a36f.png">

But at the same time we receive a 400 in the console. Maybe we have to handle the case for 'no talent found' better in the helper. Meaning that if we don't find a talent, we set the status code to something and return sth.

400 in the browser console:
![Screenshot 2021-02-25 at 17 01 05](https://user-images.githubusercontent.com/50356310/109180643-5e3abf80-778b-11eb-9558-e620c46865c9.png)

atsteinberg commented:
That is worth a shot. Although, is there a more suitable status code in this case than 400? In any case, I think this:
https://www.apollographql.com/docs/react/data/error-handling/ <https://www.apollographql.com/docs/react/data/error-handling/>
Is the place to look (especially the very last bit).
Anyway: so it seems that Apollo client is responsible for logging the error to the console. So, we somehow need to tell it to not do that in this particular case and we’re all set.
Do you agree?

> On 25. Feb 2021, at 17:04, TobiBecker <notifications@github.com> wrote:
> 
> 
> looked into it. But couldn't figure it out yet.
> The throw new ApolloError makes the error appear in the console of the server - which is fine.
>  <https://user-images.githubusercontent.com/50356310/109180490-38adb600-778b-11eb-8e95-6ab523c6a36f.png>
> But at the same time we receive a 400 in the console. Maybe we have to handle the case for 'no talent found' better in the helper. Meaning that if we don't find a talent, we set the status code to something and return sth.
> 
> 400 in the browser console:
>  <https://user-images.githubusercontent.com/50356310/109180643-5e3abf80-778b-11eb-9558-e620c46865c9.png>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/29#issuecomment-786011851>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIIRIPFQVDTYROX6OWR3TAZYHRANCNFSM4XSFD3IQ>.
>

atsteinberg commented:
I was very confused about this, since the getTalentById query was executed in a try-catch block. And even when leaving the catch block empty it still gets thrown. So I had no idea what was going on. But makes sense if it part of Apollo client’s default error handling to log to the (browser) console.

> On 25. Feb 2021, at 17:04, TobiBecker <notifications@github.com> wrote:
> 
> 
> looked into it. But couldn't figure it out yet.
> The throw new ApolloError makes the error appear in the console of the server - which is fine.
>  <https://user-images.githubusercontent.com/50356310/109180490-38adb600-778b-11eb-8e95-6ab523c6a36f.png>
> But at the same time we receive a 400 in the console. Maybe we have to handle the case for 'no talent found' better in the helper. Meaning that if we don't find a talent, we set the status code to something and return sth.
> 
> 400 in the browser console:
>  <https://user-images.githubusercontent.com/50356310/109180643-5e3abf80-778b-11eb-9558-e620c46865c9.png>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/29#issuecomment-786011851>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIIRIPFQVDTYROX6OWR3TAZYHRANCNFSM4XSFD3IQ>.
>

atsteinberg commented:
I think, the trick is to pass an options object to the useQuery with an onError key:
https://www.apollographql.com/docs/react/data/queries/#usequery-api <https://www.apollographql.com/docs/react/data/queries/#usequery-api>

If you have it open, you could just try to pass something like
{ onError:(error: ApolloError} => if (!error) console.log(’no error’) }
As options object to the relevant useQuery (shouldn’t do anything and typescript and linters don’t complain. 

> On 25. Feb 2021, at 17:04, TobiBecker <notifications@github.com> wrote:
> 
> 
> looked into it. But couldn't figure it out yet.
> The throw new ApolloError makes the error appear in the console of the server - which is fine.
>  <https://user-images.githubusercontent.com/50356310/109180490-38adb600-778b-11eb-8e95-6ab523c6a36f.png>
> But at the same time we receive a 400 in the console. Maybe we have to handle the case for 'no talent found' better in the helper. Meaning that if we don't find a talent, we set the status code to something and return sth.
> 
> 400 in the browser console:
>  <https://user-images.githubusercontent.com/50356310/109180643-5e3abf80-778b-11eb-9558-e620c46865c9.png>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/29#issuecomment-786011851>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIIRIPFQVDTYROX6OWR3TAZYHRANCNFSM4XSFD3IQ>.
>

TobiBecker committed:
revert types in getTalentById helper

beckertobias commented:
Hi Alex, it is working now.
What I did:
- added {loading, error} to the mutation options --> no error logged to browser console
- reverted the changes to getTalentById, i.e. it has to return a Talent
- checked signup again --> no error logged to browser console
- deleted the {loading, error} option --> still no error logged to console 😅 

So it is working as we want it to work. I am only not quite sure why we don't get the error in the browser console now 

Please approve if it is fine for you now.

atsteinberg commented:
Ok. Will confirm that I get the same behaviour.

> On 27. Feb 2021, at 09:39, Tobias Niklas Becker <notifications@github.com> wrote:
> 
> 
> Hi Alex, it is working now.
> What I did:
> 
> added {loading, error} to the mutation options --> no error logged to browser console
> reverted the changes to getTalentById, i.e. it has to return a Talent
> checked signup again --> no error logged to browser console
> deleted the {loading, error} option --> still no error logged to console 😅
> So it is working as we want it to work. I am only not quite sure why we don't get the error in the browser console now
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fullstacksmart/next-kara/pull/29#issuecomment-787037244>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AO7DIISEGFJOE66D3RIVRY3TBCVS5ANCNFSM4XSFD3IQ>.
>

atsteinberg commented on apollo/typedefs.ts:
```diff
@@ -307,7 +307,7 @@ const typeDefs = gql`
     getAllUserIds: [ID]!
     getAllEmployerIds: [ID]!
-    getTalentById(id: String!): Talent!
+    getTalentById(id: String!): Talent
```
could you change this back to return Talent!
?

beckertobias replied:
done

atsteinberg commented on lib/auth/firebaseAdmin.ts:
```diff
@@ -32,9 +32,13 @@ export const getUserFromToken = async (
+      console.log('found user: ', user);
+      return user;
+    } catch (e) {
+      console.error(e);
```
could you add eslint-ignore-line no-console (I think that's the right command) to this and all other console logs that will stay in the code for the time being, and remove all logs that only are there for checking purposes?

beckertobias replied:
done

atsteinberg commented on pages/signup/index.tsx:
```diff
@@ -31,13 +31,15 @@ const SignUpPage = ({ t }: PageProps): React.ReactElement => {
     name: {
       lastName: '',
     },
-    gender: Gender.OTHER,
```
I have the required changes on my computer. Will push directly to main once this pr is merged, if that's ok.

beckertobias replied:
ok, fine for me.

TobiBecker committed:
remove logs or disable eslint

beckertobias commented:
@atsteinberg did requested changes

atsteinberg reviewed (approved):
🕺
